src/raw-value.ts

```diff
@@ -126,64 +78,136 @@ function createReverseShadowTarget(target: ReverseProxyTarget): ReverseShadowTar
+function getPropertyDescriptor(o: any, p: PropertyKey): PropertyDescriptor | undefined {
+    do {
+        const d = ReflectGetOwnPropertyDescriptor(o, p);
+        if (!isUndefined(d)) {
+            ReflectSetPrototypeOf(d, null);
+            return d;
+        }
+        o = ReflectGetPrototypeOf(o);
+    } while (o !== null);
+    return undefined;
+}
+
 export function reverseProxyFactory(env: MembraneBroker) {
 
-    function getReverseDescriptor(descriptor: PropertyDescriptor): PropertyDescriptor {
-        const reverseDescriptor = assign(ObjectCreate(null), descriptor);
-        const { value, get, set } = reverseDescriptor;
-        if ('writable' in reverseDescriptor) {
+    function getRawDescriptor(secDesc: PropertyDescriptor): PropertyDescriptor {
+        const rawDesc = assign(ObjectCreate(null), secDesc);
+        const { value, get, set } = rawDesc;
+        if ('writable' in rawDesc) {
             // we are dealing with a value descriptor
-            reverseDescriptor.value = isFunction(value) ?
+            rawDesc.value = isFunction(value) ?
                 // we are dealing with a method (optimization)
                 getRawFunction(value) : getRawValue(value);
         } else {
             // we are dealing with accessors
             if (isFunction(set)) {
-                reverseDescriptor.set = getRawFunction(set);
+                rawDesc.set = getRawFunction(set);
             }
             if (isFunction(get)) {
-                reverseDescriptor.get = getRawFunction(get);
+                rawDesc.get = getRawFunction(get);
             }
         }
-        return reverseDescriptor;
+        return rawDesc;
     }
 
-    function copyReverseOwnDescriptors(shadowTarget: ReverseShadowTarget, descriptors: PropertyDescriptorMap) {
-        for (const key in descriptors) {
-            // avoid poisoning by checking own properties from descriptors
-            if (hasOwnProperty(descriptors, key)) {
-                const originalDescriptor = getReverseDescriptor(descriptors[key]);
-                installDescriptorIntoShadowTarget(shadowTarget, key, originalDescriptor);
+    function getSecurePartialDescriptor(rawPartialDesc: PropertyDescriptor): PropertyDescriptor {
+        const secPartialDesc = assign(ObjectCreate(null), rawPartialDesc);
+        if ('value' in secPartialDesc) {
+            // we are dealing with a value descriptor
+            secPartialDesc.value = env.getSecureValue(secPartialDesc.value);
+        }
+        if ('set' in secPartialDesc) {
+            // we are dealing with accessors
+            secPartialDesc.set = env.getSecureValue(secPartialDesc.set);
+        }
+        if ('get' in secPartialDesc) {
+            secPartialDesc.get = env.getSecureValue(secPartialDesc.get);
+        }
+        return secPartialDesc;
+    }
+
+    function lockShadowTarget(shadowTarget: ReverseShadowTarget, originalTarget: ReverseProxyTarget) {
```

Comment:
☝️ the `lockShadowTarget` should check if the `shadowTarget` is already locked before locking it again.

Reply:
we either check for that in the callers of `lockShadowTarget()` before calling it, or we check here and exit.